Add release notes for the 0.9.0 Metals version (#1689)
scalameta/metals: website/blog/2020-05-04-lithium.md

---
author: Tomasz Godzik
title: Metals v0.9.0 - Lithium
authorURL: https://twitter.com/TomekGodzik
authorImageURL: https://github.com/tgodzik.png
---

We are happy to announce the release of Metals v0.9.0, codename "Lithium", which
main focus is introducing support for Scala 3 starting with versions 2.13.0 and

> 2.14.0-RC1. We also added support for Scala 2.13.2 and a number of other useful

Reviewer: What is version 2.14.0-RC1?

---
author: Tomasz Godzik
title: Metals v0.9.0 - Lithium
authorURL: https://twitter.com/TomekGodzik
authorImageURL: https://github.com/tgodzik.png
---

We are happy to announce the release of Metals v0.9.0 codename "Lithium", which
main focus is introducing the support for Scala 3 starting with the versions

> 2.13.0 and 2.14.0-RC1. We also added support for Scala 2.13.2 and a number of

Author: :fearful:

> 2.13.0 and 2.14.0-RC1. We also added support for Scala 2.13.2 and a number of

Author: Too many 2s and 3s :sweat_smile:  Thanks for noticing!

> 2.13.0 and 2.14.0-RC1. We also added support for Scala 2.13.2 and a number of

Reviewer: I was concerned I had missed some important announcement!

That said, there are some issues that are lacking in the Scala 3 support that
are available for 2.x versions. Most notably, we are still missing all of the
more advanced completions, filtering and sorting of completion items or auto
imports. There are also some possible issues connected to optional braces and
top level methods features. In the upcoming months we are planning to address
most of those issues and implement the lacking features.

You can find missing features in the
[feature requests repository](https://github.com/scalameta/metals-feature-requests/issues?q=is%3Aissue+is%3Aopen+label%3Ascala3)
and currently known bugs in the
[main Metals repo.](https://github.com/scalameta/metals/issues?q=is%3Aissue+is%3Aopen+label%3A%22Scala+3%22)

The first supported versions will be `0.23.0` and `0.24.0-RC1`, which are the
latest stable and latests release candidate respectably. We plan to always
support such two versions until there is a more stable release cycle for
Scala 3.

Please do report any issues you encounter, especially connected to some new
Scala 3 features, which will be an immense help to the team.

## Improvements related to Bloop/Metals integration

> There has been multiple reported issues in the last weeks about hanging
> compilation, which could make the Metals workspace hard to use. Our top priority
> is currently figuring out all the possible issues both in Metals as well in
> Bloop. Thanks to some fixes done by [jvican](https://github.com/jvican) we are
> hopeful that some of the issues might become less frequent. Those fixes are
> available in the current Bloop snapshots and in the version included in Metals

Reviewer 2: ```suggestion
There have been multiple reported issues in the last weeks about hanging
compilation, which can make the Metals workspace hard to use. Our top priority
has been figuring out all the possible issues both in Metals as well as in
Bloop. Thanks to some fixes done by [jvican](https://github.com/jvican), and some
work on the Metals side on the Bloop integration we're hopeful that some of the
issues are becoming less frequent. These fixes are included in this release.
```
I think we should re-word this a bit to sound a bit more positive. What do you think of this? I also cut out some of the details that are sort of imo irrelevant to this release doc like "these fixes are available in the current Bloop snapshots. I think this is a bit more positive and a bit more to the point.

Author: Thanks, I applied these changes.

We also reworked the Metals tree view to highlight some commands that a user can
use to fix their workspace as well as added a new command to recompile workspace
from scratch:

![reworked-tree-view](https://i.imgur.com/PERfNqt.png)

`Clean compile workspace` might be needed in case of broken Zinc caches and
`Restart Bloop server` should help with the cases where deduplication of
compilation would hang. Do not hesitate to report any possible issues and if
possible there are a number of additional diagnostics that can be done before

> reporting an issue:

Reviewer 2: Tbh I know this is extremely helpful to have for an issue, but I'm feeling a bit like these release notes are starting to sway pretty heavy on the tech side rather than an overview of the features and functionality this release has to offer. Again, I totally think we should include some of the technical improvements, but I'm not sure how I feel about telling users to "Give us a thread dump of the Bloop process" here. Imagine a person new to Scala just getting started with Metals reading over these release notes? Thoughts?

Author: Maybe I went over the board :sweat_smile: . I will reduce it to "please report an issue". We can guide the user later. Though it would be extremely useful have that reported right from the bat, I spent loads of time on digging and didn't manage to get anywhere. However, I agree it might be too much.

Author: Changed the section a bit and now it's just:
`Even if those commands help, do not hesitate to report any issues. We are also
monitoring the situation ourselves and investigating all problems encountered.`

Reviewer 2: Yea I agree it's important. Maybe we should re-work the issue template to say that if they are dealing with this specific type of issue, these are the things that would be helpful.

Author: Maybe it will not be necessary. I haven't noticed the issue since Bloop updated, so here's hoping! :crossed_fingers:

> - getting a thread dump of Bloop process

Reviewer 2: ```suggestion
- Getting a thread dump of the Bloop process
```

> - running verbose compilation via the CLI:

Reviewer 2: ```suggestion
- Running verbose compilation via the CLI:
```

  `coursier launch ch.epfl.scala:<bloop_version> -- compile --verbose <project_name>`

> - (advanced) attaching profiler such as YourKit to discover exceptions being

Reviewer 2: ```suggestion
- (Advanced) attaching a profiler such as YourKit to discover exceptions being
```

>   thrown by the Bloop server. This will require

Reviewer 2: I think you cut off in the middle of the sentence here.

We are also monitoring the situation ourselves and investigating all problems
encountered.

## Automatically add '+' on newline inside a string

The Metals team was recently joined by [mlachkar](https://github.com/mlachkar),

> an engineer from Scala Center, who already contributed a number of features and

Reviewer 2: ```suggestion
an engineer from the Scala Center, who already contributed a number of features and
```

> bug fixes to the Project. Thanks to her, we now can easily enter a newline

Reviewer 2: ```suggestion
bug fixes to the Project. Thanks to her, we can now easily enter a newline
```

inside a single line string and that string will be correctly split with an
additional `+`.

![split-line](https://i.imgur.com/uhF0MOx.gif)

## Code action to import all missing symbols

Previously, it was only available to import one symbol at a time in case of
missing imports. Thanks to [colineto](https://github.com/colineto) we now have

> an ability to import everything in bulk for every unambiguous import available.

Reviewer 2: ```suggestion
the ability to import everything in bulk for every unambiguous import available.
```

Unambiguous in this case meaning that there is only one possible import to
choose from.

![import-all](https://i.imgur.com/mmzgJs7.gif)

## Miscellaneous improvements

- added an option to ignore .scalafmt.conf creation message
- fixed hanging while debugging if workspace had errors
- fixed issue with folding in non 2.12.11 workspaces
- turn off "Go to parent" code lenses by default
- fixed false parsing errors in 2.13.x workspaces
- added date string to metals logs
- fixed issue with additional src directories being created automatically

> - fixed "Go to implementation" to work with local aliases

Reviewer 2: Capitalize the first letter of these

Reviewer 2: You could also maybe add "Fixed auto-import issue in package objects"

Author: Forgot about that one! I remembered to add to PR list, but then my mind wandered.

## Contributors

Big thanks to everybody who contributed to this release! We had a lot of new
contributors doing some really exemplary work.

```
$ git shortlog -sn --no-merges v0.8.4..v0.9.0
Ayoub Benali
Tomasz Godzik
Scala Steward
Chris Kipp

> Coline

Author: @colineto Do you want to be fully mentioned here with the surname or should I leave it as is?

Reviewer 3: Coline Thomas will be great thx

Author: Updated!
